File: mcmc/matlab/mcmc.m
% Target probability distribution: a mixture of two Gaussian distributions
function p = targetDistribution(x)
    gaussian1 = 0.3 * exp(-0.2 * (x - 10).^2);
    gaussian2 = 0.7 * exp(-0.2 * (x + 10).^2);
    p = gaussian1 + gaussian2;
end

% Metropolis-Hastings algorithm for MCMC sampling
function samples = metropolisHastings(target, numSamples, initialState, proposalStd)
    currentState = initialState;
    accepted = 0;
    samples = zeros(1, numSamples);
    
    for i = 1:numSamples
        % Generate a proposal state from a normal distribution
        proposal = currentState + proposalStd * randn();
        
        % Calculate the acceptance probability
        acceptanceProb = min(1, target(proposal) / target(currentState));
        
        % Accept or reject the proposal
        if rand() < acceptanceProb
            currentState = proposal;
            accepted = accepted + 1;
        end
        
        % Save the current state as a sample
        samples(i) = currentState;
    end
    
    acceptanceRate = accepted / numSamples;
    fprintf('Acceptance rate: %.4f\n', acceptanceRate);
end

% Define parameters for MCMC sampling
numSamples = 10000;
initialState = 0.0;
proposalStd = 5.0;

% Run MCMC sampling using Metropolis-Hastings algorithm
samples = metropolisHastings(@targetDistribution, numSamples, initialState, proposalStd);

% Plot the sampled distribution
histogram(samples, 'Normalization', 'pdf');
xlabel('x');
ylabel('Probability Density');
title('Sampled Distribution');

File: mcmc/python/mcmc.py
import numpy as np
import matplotlib.pyplot as plt

# Target probability distribution: a mixture of two Gaussian distributions
def target_distribution(x):
    return 0.3 * np.exp(-0.2 * (x - 10)**2) + 0.7 * np.exp(-0.2 * (x + 10)**2)

# Metropolis-Hastings algorithm for MCMC sampling
def metropolis_hastings(target, num_samples, initial_state, proposal_std):
    samples = np.zeros(num_samples)
    current_state = initial_state
    accepted = 0

    for i in range(num_samples):
        # Generate a proposal state from a normal distribution
        proposal = np.random.normal(current_state, proposal_std)

        # Calculate the acceptance probability
        acceptance_prob = min(1, target(proposal) / target(current_state))

        # Accept or reject the proposal
        if np.random.rand() < acceptance_prob:
            current_state = proposal
            accepted += 1

        # Save the current state as a sample
        samples[i] = current_state

    acceptance_rate = accepted / num_samples
    print("Acceptance rate:", acceptance_rate)
    return samples

# Set random seed for reproducibility
np.random.seed(42)

# Define parameters for MCMC sampling
num_samples = 10000
initial_state = 0
proposal_std = 5

# Run MCMC sampling using Metropolis-Hastings algorithm
samples = metropolis_hastings(target_distribution, num_samples, initial_state, proposal_std)

# Plot the sampled distribution against the target distribution
x = np.linspace(-20, 20, 1000)
plt.plot(x, target_distribution(x), label="Target Distribution")
plt.hist(samples, bins=50, density=True, alpha=0.5, label="Sampled Distribution")
plt.xlabel("x")
plt.ylabel("Probability Density")
plt.legend()
plt.show()
